A more complete manual

panglesd opened the pull request:
This PR modifies and complete the current manual.
Since it is bigger, it is split in several pages.
Some parts in "good practices" are not yet written, but I believe the manual is already large enough to be merged, and it can be completed in later PRs, either by me or other contributors.
The manual can be seen rendered in https://choum.net/panglesd/ppxlib_new_manual/ppxlib/index.html

I think this is a good step toward #324! Although there are still a lot of improvements that could be done.

panglesd commented:
I will squash all commits before merging, this should be reviewed as a whole.

avsm commented on doc/quick_intro.mld:
```diff
@@ -0,0 +1,79 @@
+{1 PPXs and Ppxlib history}
+
+The preprocessing history of OCaml started long before Ppxlib. Several tools
+have occupied the place that Ppxlib occupies now, TODO
```
not worth spending too much time on this, but the obvious ones are camlp4 and camlp5.  The original PPX discussions are archived here: https://lists.ocaml.org/pipermail/wg-camlp4/2013-January/000000.html

panglesd replied:
I added some links and a link to contribute to the doc (I don't really have much more time to write more sections, I'll focus on getting this merged. Missing sections would be added later)

avsm commented on doc/ast-traversal.mld:
```diff
@@ -0,0 +1,113 @@
+subvalues for a certain type. For instance, [List.map] is a traversal of the
+[list] type. In the case of a [list], a map is very simple to write, but in the
+case of the very complex [Parsetree] type, it is a lot of code! Fortunately,
+Ppxlib provides way to ease this.
```
a style nit: "Ppxlib" renders very badly. Perhaps you should refer to it consistently as ppxlib rather than capitalising it?

panglesd replied:
You are right! I checked how it was added [here](https://github.com/ocaml/ocaml.org/pull/535), so I replaced every occurrence of Ppxlib to `ppxlib`.

avsm commented on doc/ast-traversal.mld:
```diff
@@ -0,0 +1,113 @@
+of Parsetree manipulations, known as traversals.
+
+A traversal is a recursive function which will be called on a value and all
+subvalues for a certain type. For instance, [List.map] is a traversal of the
```
```suggestion
subvalues for a certain type. For instance, {!List.map} is a traversal of the
```

avsm commented on doc/ast-traversal.mld:
```diff
@@ -0,0 +1,113 @@
+  replace_constant#payload payload
+]}
+
+Note that when redefining methods, unless explicitely wanting the traversal to
```
```suggestion
Note that when redefining methods, unless explicitly wanting the traversal to
```

avsm commented on doc/examples.mld:
```diff
@@ -0,0 +1,159 @@
+  | value -> Ast_builder.Default.estring ~loc value
+  | exception Not_found ->
+      let ext =
+        Location.error_extensionf ~loc "The environement variable %s is unbound"
```
```suggestion
        Location.error_extensionf ~loc "The environment variable %s is unbound"
```

avsm commented on doc/generating-code.mld:
```diff
@@ -0,0 +1,336 @@
+
+However, sometimes you might want to use a feature that was introduced recently
+in OCaml, and is not integrated in {!Ppxlib.Ast_builder}. For instance, OCaml
+4.14 introduced the possibility to explicitely introduce type variable in
```
```suggestion
4.14 introduced the possibility to explicitly introduce type variables in a
```

avsm commented on doc/generating-code.mld:
```diff
@@ -0,0 +1,336 @@
+However, sometimes you might want to use a feature that was introduced recently
+in OCaml, and is not integrated in {!Ppxlib.Ast_builder}. For instance, OCaml
+4.14 introduced the possibility to explicitely introduce type variable in
+constructor declaration. This modified sligthly the AST type, and for
```
```suggestion
constructor declaration. This modified the AST type, and for
```

avsm commented on doc/generating-code.mld:
```diff
@@ -0,0 +1,336 @@
+in OCaml, and is not integrated in {!Ppxlib.Ast_builder}. For instance, OCaml
+4.14 introduced the possibility to explicitely introduce type variable in
+constructor declaration. This modified sligthly the AST type, and for
+retro-compatibility, {!Ppxlib.Ast_builder} did not modify the signature of the
```
```suggestion
backwards compatibility, {!Ppxlib.Ast_builder} did not modify the signature of the
```

avsm reviewed (commented):
This is a _really_ useful contribution to the ppx ecosystem, thanks @panglesd!  I had a quick read through and left some comments, but overall I think this should definitely ship and be publicised for further feedback.  Nice work!  The only thing I could think of adding is a section on some good example PPXes that illustrate the various sections with links to them.

panglesd commented:
Thanks a lot for your review comments and encouragements!

>  The only thing I could think of adding is a section on some good example PPXes that illustrate the various sections with links to them.

I think the [Example](https://choum.net/panglesd/ppxlib_new_manual/ppxlib/examples.html) section should be used for this, updating the `examples` folder at the same time (later on, I will use mdx to check code blocks, and import examples from their files to avoid duplicating them).
In my opinion, all of these improvements should definitely be made, but in another PR: I've delayed opening this PR for a long time, wanting to finish or add or polish many parts, but in the end all the things already written are not available to the public, which is a shame!

avsm commented:
I entirely agree you can do improvements in a followup PR. What's here is an excellent start and very shippable.

panglesd commented:
@pitag-ha @davesnx @christinerose I finished a first pass of auto-reviewing, and updated https://choum.net/panglesd/ppxlib_new_manual/ppxlib/index.html with the new version.

pitag-ha commented:
Wow @panglesd, this really looks amazing! (I feel very lucky to come back from vacation, and among the first things I see is this great PR)

From what I understand, @christinerose is currently reviewing for language, right? So I'll wait for her to finish and afterward, I'll do a quick review myself for content. But I'm with you: let's merge this pretty much as is. We can always improve things and make it more complete later.

christinerose reviewed (changes requested):
Mostly grammar (commas), tightening some syntax, caught some typos, corrected a few verb agreements, and left a couple of comments/questions for clarification.

christinerose reviewed (changes requested):
Mostly grammar (commas), tightening some syntax, caught some typos, corrected a few verb agreements, and left a couple of comments/questions for clarification.

panglesd commented on doc/ast-traversal.mld:
```diff
@@ -0,0 +1,111 @@
+For each kind of traversal described above, [ppxlib] provides "default" traversal,
+in the form of a class. For instance, in the case of the map traversal, the
+default map is the identity, and any object of class {{!Ppxlib.Ast_traverse.map}[Ast_traverse.map]}
+will be the identity. To apply this identity to a node of a given type, one need
```
I will rewrite this sentence a bit to make it clearer that all of this is the same identity!

panglesd commented on doc/driver.mld:
```diff
@@ -0,0 +1,377 @@
+author, it provides an API to define the transformation, and register it to
+[ppxlib]. Then, all registered transformations can be turned into a single
+executable, called the {e driver}, that is responsible for applying all the
+transformations, and that will be called by the compiler.
```
Auto-suggestions (it's not the transformation that will be called by the compiler):
```suggestion
transformations. The driver will be called by the compiler (or, in standalone mode).
```

christinerose replied:
That's perfect!

panglesd commented on doc/driver.mld:
```diff
@@ -0,0 +1,377 @@
+{1:derivers-and-extenders Context-free transformations}
+
+[ppxlib] defines several kinds of transformations whose core property is that they
+can only read and modify the code locally: the parts of the AST that are given
```
```suggestion
can only read and modify the code locally. The parts of the AST given
```

panglesd commented on doc/driver.mld:
```diff
@@ -0,0 +1,377 @@
+general-purpose as plain AST transformations, they are more than often
+sufficient and have many nice properties such as a well-defined semantic for
+composition.
+The two most important of such context-free transformations are {e derivers} and
```
```suggestion
The two most important context-free transformations are {e derivers} and
```

panglesd commented on doc/driver.mld:
```diff
@@ -0,0 +1,377 @@
+A {e deriver} is a context-free transformation that, given a structure or
+signature item, will generate code {e to append after} this item. The given code is
+never modified. A deriver can be very useful to generate values depending on the
+structure of a user-defined type, for instance a converter of a value of a type
```
```suggestion
structure of a user-defined type, for instance a converter for a type
```

panglesd commented on doc/driver.mld:
```diff
@@ -0,0 +1,377 @@
+
+{2 Advantages}
+
+The advantages of using context-free transformations are multiple. First, they
+provide to the user a much clearer understanding of the parts of the AST that
+will be rewritten, than a fully general AST rewriting. Secondly, they provide
+a much better composition semantic, which does not depend on the order. Finally,
+context-free transformation are applied in a single phase factorising the work
+for all transformation, resulting in a much faster driver than when combining
+multiple whole AST transformations. More details on the execution of this phase
+are given in the {{!"context-free-phase"}dedicated section.}
```
I add back that the advantage is for the PPX user (as opposed to the PPX writer)
```suggestion
There are multiple advantages of using context-free transformations. First, they
provide the PPX user a much clearer understanding of the AST parts that
will be rewritten, rather than a fully general AST rewriting. Secondly, they provide
a much better composition semantic, which does not depend on the order. Finally,
context-free transformations are applied in a single phase factorising the work
for all transformations, resulting in a much faster driver than when combining
multiple, whole AST transformations. More details on the execution of this phase
are given in its {{!"context-free-phase"}dedicated section.}
```

panglesd commented on doc/driver.mld:
```diff
@@ -0,0 +1,377 @@
+{2:global-transfo-phase The global transformation phase}
+
+The global transformation phase is the phase where registered transformations,
+seen as function from and to the {{!Ppxlib.Parsetree}[Parsetree]}, are run. The order on which they are
+applied might matter and change the outcome, but since [ppxlib] knows nothing
+about the transformations, the order applied is alphabetical on the names of the
+transformation.
```
typo in "transormation" :)
```suggestion
seen as function from and to the {{!Ppxlib.Parsetree}[Parsetree]}, are run. The applied order might matter and change the outcome, but since [ppxlib] knows nothing
about the transformations, the order applied is alphabetical by the transformation's name.
```

panglesd commented on doc/driver.mld:
```diff
@@ -0,0 +1,377 @@
+The {{!https://dune.build}Dune} build system is well integrated with the [ppxlib]
+mechanism of registering transformations. Dune will read in every [dune] file
+the set of PPXs that are to be used. For a given set of rewriters, it will
+generate a driver using {{!Ppxlib.Driver.run_as_ppx_rewriter}[Driver.run_as_ppx_rewriter]} containing the registered
+transformation of the whole set. Using a single driver for multiple
+transformations from multiple PPXs ensure a better composition semantics and
+improves the speed of the combined transformations.
+
+Moreover, [ppxlib] communicate with Dune through [.corrected] files, to allow for
+promotion. Inlining rewriters generate such files, but the PPX author can
+generate its own promotion suggestion using the
+{{!Ppxlib.Driver.register_correction}[Driver.register_correction]} function.
```
```suggestion
generate a driver using {{!Ppxlib.Driver.run_as_ppx_rewriter}[Driver.run_as_ppx_rewriter]} that contains all registered
transformation. Using a single driver for multiple
transformations from multiple PPXs ensures better composition semantics and
improves the speed of the combined transformations.
Moreover, [ppxlib] communicates with Dune through [.corrected] files to allow for
promotion. Inlining rewriters generate such files, but the PPX author can
generate its own promotion suggestion using the
{{!Ppxlib.Driver.register_correction}[Driver.register_correction]} function.
```

panglesd commented on doc/generating-code.mld:
```diff
@@ -0,0 +1,357 @@
+{2 General presentation}
+
+The {{!Ppxlib.Ast_builder}[Ast_builder]} module provides several kinds of functions to generate AST
+nodes. The first kind are ones whose name matches closed the {{!Ppxlib.Parsetree}[Parsetree]}
```
I think it's supposed to be:
```suggestion
nodes. The first kind are ones whose name matches closely the {{!Ppxlib.Parsetree}[Parsetree]} type names.
```

panglesd commented on doc/generating-code.mld:
```diff
@@ -0,0 +1,357 @@
+{2 General presentation}
+
+As you have seen, defining code with {{!Ppxlib.Ast_builder}[Ast_builder]} does
+not feel perfectly natural: some knowledge of the [Parsetree] types is needed.
```
```suggestion
not feel perfectly natural. Some knowledge of the [Parsetree] types is needed.
```

panglesd commented on doc/generating-code.mld:
```diff
@@ -0,0 +1,357 @@
+
+Note that if an anti-quote extension node is in the wrong context, it won't be
+rewritten by {{!Ppxlib_metaquot}[Metaquot]}. For instance, in [[%expr match [] with [%e some_value] -> 1]]
+the anti-quote extension node for expression is put in a pattern context,
```
It is not for a particular expression, but for expressions in general. Maybe:
```suggestion
the anti-quote extension node for expressions is put in a pattern context,
```

panglesd commented on doc/good-practices.mld:
```diff
@@ -0,0 +1,402 @@
+to report errors, instead of the more merlin friendly way consisting of
+embedding errors in the AST, as described in this section.
+
+If you want to migrate such a codebase to the embedding approach, here are a few
+recipes to do that. Indeed, it might not be completely trivial, as raising can
+be done anywhere in the code, including in places where "embedding" would not
+make sense. What you can do is to turn your internal raising functions to
+function returning a [result] type.
```
```suggestion
If you want to migrate such a codebase to the embedding approach, the rest of this section will present few
recipes to do that. It might not be completely trivial, as raising can
be done anywhere in the code, including in places where "embedding" would not
make sense. The first thing you can do is to turn your internal raising functions to
function returning a [result] type.
```

panglesd commented on doc/good-practices.mld:
```diff
@@ -0,0 +1,402 @@
+The workflow for this change would look like this:
+
+{ol
+{- Search through your code all uses of {{!Ppxlib.Location.raise_errorf}[raise_errorf]}, using for instance [grep].}
+{- For each of them, turn them into function returning a [(_, extension) result] type, using {{!Ppxlib.Location.error_extensionf}[error_extensionf]} to generate the [Error].}
+{- Let the compiler or merlin tell you where you need to propagate the [result] type (most certainly using [map]s and [bind]s).}
```
Misplaced `}`
```suggestion
{- Search your code for all uses of {{!Ppxlib.Location.raise_errorf}[raise_errorf]}, using [grep], for instance.}
{- For each of them, turn them into functions returning a [(_, extension) result] type, using {{!Ppxlib.Location.error_extensionf}[error_extensionf]} to generate the [Error].}
{- Let the compiler or Merlin tell you where to propagate the [result] type (most certainly using [map]s and [bind]s).}
```

panglesd commented on doc/good-practices.mld:
```diff
@@ -0,0 +1,402 @@
+{- Search through your code all uses of {{!Ppxlib.Location.raise_errorf}[raise_errorf]}, using for instance [grep].}
+{- For each of them, turn them into function returning a [(_, extension) result] type, using {{!Ppxlib.Location.error_extensionf}[error_extensionf]} to generate the [Error].}
+{- Let the compiler or merlin tell you where you need to propagate the [result] type (most certainly using [map]s and [bind]s).}
+{- When you have propagated until a point where you can, embed the extension in case of [Error extension].}
```
```suggestion
{- When you have propagated until a point where you can embed an extension node, turn the [Error] case into an extension node and embed it.}
```

panglesd commented on doc/writing-ppxs.mld:
```diff
@@ -0,0 +1,512 @@
+
+- It is harder for the user to know exactly what parts of the AST will be
+  changed.
+- It is harder for [ppxlib] to combine several global transformations: there is no
+  guarantee that the effect of one will work well with the effect of another.
+- The job done by two global transformations (e.g. an AST traverse) cannot be
+  factorized, resulting in slower compilation time.
+
+For all these reasons, a global transformation should be avoided whenever a
+context-free transformation could do the job, which by experience seems to be most of the time.
+
+The API for defining a global transformation is easy: a global transformation consists simply of the function, and can directly be registered with {{!Ppxlib.Driver.register_transformation}[Driver.register_transformation]}.
```
Misplaced space:
```suggestion
- It is harder for [ppxlib] to combine several global transformations, as there is no
  guarantee that the effect of one will work well with the effect of another.
- The job done by two global transformations (e.g., an AST traverse) cannot be
  factorised, resulting in slower compilation time.
For all these reasons, a global transformation should be avoided whenever a
context-free transformation could do the job, which by experience seems to be most of the time.
The API for defining a global transformation is easy. A global transformation consists simply of the function and can be directly be registered with {{!Ppxlib.Driver.register_transformation}[Driver.register_transformation]}.
```

panglesd commented on doc/driver.mld:
```diff
@@ -0,0 +1,377 @@
+
+{2 The driver}
+
+[ppxlib] sits in between the PPXs authors, and the compiler toolchain. For the PPX
+author, it provides an API to define the transformation, and register it to
+[ppxlib]. Then, all registered transformations can be turned into a single
+executable, called the {e driver}, that is responsible for applying all the
+transformations, and that will be called by the compiler.
```
```suggestion
[ppxlib] sits in between the PPXs authors and the compiler toolchain. For the PPX
author, it provides an API to define the transformation and register it to
[ppxlib]. Then, all registered transformations can be turned into a single
executable, called the {e driver}, that is responsible for applying all the
transformations. The driver will be called by the compiler.
```

panglesd commented:
Thanks a lot @christinerose for the impressively careful review, github seems to struggle when there are so many comments! I merged most of them, sometimes with small modifications (answering with an "auto-suggestion")

I'll squash many of the commits before merging!

christinerose commented:
> Thanks a lot @christinerose for the impressively careful review, github seems to struggle when there are so many comments! I merged most of them, sometimes with small modifications (answering with an "auto-suggestion")
> 
> I'll squash many of the commits before merging!

Thank you! I'll definitely do them in smaller chunks next time, directly editing small grammar corrections to save so many suggestions. 🙂

panglesd commented:
I included some improvements on the API landing page. I reordered the modules, added sections and descriptions, and moved `Quoter` into `Expansion_helpers`, as mentioned in #370. @pitag-ha could you confirm that the API changes are OK?

Finally, I added links from the API to the manual, in relevant places.

davesnx commented on doc/driver.mld:
```diff
@@ -0,0 +1,372 @@
+phase}, registering to this phase provides no guarantee that the transformation
+is run early in the rewriting, as there is no limit in the number of
+transformations registered in this phase, which are then applied in the
+alphabetical order by their name.
```
TIL "alphabetical order by their name"

davesnx commented on doc/generating-code.mld:
```diff
@@ -0,0 +1,357 @@
+
+{0 Generating AST Nodes}
+
+The rewriter's core is a function that outputs code in the form of an AST.
```
Unsure where, but I would add a link to astexplorer. If you are a first-timer here you would have a easier understanding once you see the parsetree as a tree and all the info there.

panglesd replied:
Do you think that's necessary? (it is already mentioned in https://ocaml.org/docs/metaprogramming#ocamls-parsetree-the-ocaml-ast which is supposed to be a "prerequisite". The manual is targeted to authors of PPXs, who should have the idea of the AST as a tree.)

I could be added, but I feel it fits better in general explanations about AST/tutorial/good practices/recipes than in the manual. Maybe we can leave this improvement/new section for later?

davesnx replied:
Yes, totally both options are great here. Another section to get deep into Parsetree and assuming readers to be aware of it works great

davesnx commented on doc/writing-ppxs.mld:
```diff
@@ -0,0 +1,510 @@
+
+{2 Extenders}
+
+An {{!driver.def_extenders}extender} is characterised by several things:
```
This reference seems broken in https://choum.net/panglesd/ppxlib_new_manual/ppxlib/writing-ppxs.html#extenders

panglesd replied:
yes, I did not kept everything up to date as I made changes too often, but you can have a look at the up to date doc by doing
```
$ dune build @doc 
$ firefox _build/default/_doc/_html/index.html
```
(if you are using firefox :wink:)

davesnx commented on doc/writing-ppxs.mld:
```diff
@@ -0,0 +1,510 @@
+
+An {{!driver.def_extenders}extender} is characterised by several things:
+
+- The context on which it applies. Indeed, extension points can be used in
```
I find this bullet a little confusing, mostly the end.
What about rewriting the sentence with a little more specific terms?

panglesd replied:
You are right, I tried to clarify!

davesnx commented on doc/generating-code.mld:
```diff
@@ -0,0 +1,357 @@
+structure items, expressions, patterns, etc., but it is not possible to generate a
+value of type {{!Ppxlib.Parsetree.row_field_desc}[row_field_desc]}!
+
+Note: The OCaml compiler API also provides a module, called {{!Ppxlib.Ast_helper}[Ast_helper]}, to
```
I would either remove this or explain that it's preferred to use Ast_builder when you need fine granted Nodes

panglesd replied:
I am not sure about the status between Ast_helper and Ast_builder, which is why I was vague but I still wanted to mention it since it is in the API. As suggested, I recommended Ast_builder.

panglesd commented on doc/driver.mld:
```diff
@@ -0,0 +1,372 @@
+
+The {{!Ppxlib.Driver.standalone}[Driver.standalone]} function creates an executable that
+parses an OCaml file, transforms it according to the registered transformations,
+and outputs the transformed file. This makes it suitable for use with the [-pp]
```
Sure, thanks for the comment. I chose the manual for 5.0, it gives less the feeling of outdated docs!

panglesd commented:
@pitag-ha @davesnx I have made a quick [experiment](https://choum.net/panglesd/ppxlib_one_page_manual/_html/ppxlib/manual.html).

Do you think a one page manual is better? I like the global navigation bar (although it is a bit too "populated"...).

davesnx commented:
I love it, but share the same concern about "populated", maybe it's a great addition and we could improve the UI of the sidebar with a collapsible?

panglesd commented:
Although the UI improvement is definitely planned, it is not on ppxlib's side. It is an odoc issue (and/or ocaml.org) which should be solved in few month!
I think the idea is to have the navbar populated with collapsible from all doc pages.

So the question is: In the meantime it is solved upstream, is it better to have the navbar "not populated enough" (as in the current PR) or "too populated" (as in the experiment)?

I would tend to prefer the "too populated" as it also removes the need for custom inserted html (aka odoc hacks).

pitag-ha commented on doc/good-practices.mld:
```diff
@@ -0,0 +1,405 @@
+
+{1 Testing Your PPX}
+
+This section is not yet written. You can refer to {{:https://tarides.com/blog/2019-05-09-an-introduction-to-ocaml-ppx-ecosystem#testing-your-ppx}this blog post}, or contribute to the [ppxlib] documentation by opening a pull request in the {{:https://github.com/ocaml-ppx/ppxlib/}repository}.
```
```suggestion
This section is not yet written. You can refer to {{:https://tarides.com/blog/2019-05-09-an-introduction-to-ocaml-ppx-ecosystem#testing-your-ppx}this blog post} (notice that that blog post was written before `dune` introduced its cram test feature), or contribute to the [ppxlib] documentation by opening a pull request in the {{:https://github.com/ocaml-ppx/ppxlib/}repository}.
```

pitag-ha commented on doc/good-practices.mld:
```diff
@@ -0,0 +1,405 @@
+
+This section is not yet written. You can contribute to the [ppxlib] documentation by opening a pull request in the {{:https://github.com/ocaml-ppx/ppxlib/}repository}.
+
+{1 Migrate From [Ppx_traverse]}
+
+This section is not yet written. You can contribute to the [ppxlib] documentation by opening a pull request in the {{:https://github.com/ocaml-ppx/ppxlib/}repository}.
```
Oh, interesting! You're mentioning Ppx_traverse and not mentioning Camlp4/Camlp5. Where have you come across Ppx_traverse?

pitag-ha replied:
The short history in this context is:

`ppx_traverse` is one of the preprocessing libraries that was used at Jane Street before `ppxlib`. Jane Street has fully shifted to `ppxlib` in everything preprocessing related.

Parallel to Jane Street's development, the OCaml community has shifted from Camlp4/Camlp5 (i.e. extending the OCaml parser itself) over writing compiler version specific PPXs just with `compiler-libs`, over writing cross-compiler compatible PPXs with `compiler-libs`+`OMP1`, to writing the PPXs we know today with `ppxlib`.

To be honest, I'd remove `Ppx_traverse`.

We _could_ mention Camlp5/Camlp4 and/or `cppo`. According to the [OCaml 2022 user survey](https://ocaml-sf.org/docs/2022/ocaml-user-survey-2022.pdf), 6,6% of people use a cpp-style conditional-compilation directives for preprocessing in OCaml and around 5% still use Camlp4/5.

ceastlund replied:
`ppx_traverse` is not an alternative to `ppxlib`, it's a `deriving` clause usable via `ppxlib`. It's used to generate the code for `ast_traverse.ml`, for example, and we use it for other things at Jane Street.

pitag-ha replied:
Thanks for the context, @ceastlund! It's good to know that it's still used. The reason why I thought `ppxlib` comes from `ppx_traverse` (among many others) is the following sentence in our [history md](https://github.com/ocaml-ppx/ppxlib/blob/main/HISTORY.md):

> This repository was originally a merge of several ppx-related repositories, namely:
>  -   [ppx_ast](https://github.com/janestreet/ppx_ast);
> -  [ppx_core](https://github.com/janestreet/ppx_core);
>  -  [ppx_driver](https://github.com/janestreet/ppx_driver);
>  -  [ppx_metaquot](https://github.com/janestreet/ppx_metaquot);
>  -  [ppx_traverse](https://github.com/janestreet/ppx_traverse);
>  -  [ppx_traverse_builtins](https://github.com/janestreet/ppx_traverse_builtins);
>  -  [ppx_type_conv](https://github.com/janestreet/ppx_type_conv).

If you could give us more context about this, I think that would be very helpful for all of us!

pitag-ha reviewed (commented):
Very nice!

I've already read quickly through the first few sections and have left a few comments. So far, I think both the newly written parts are very well-written and helpful, and the re-assembling of the already existing bits that were scattered around different places makes things far easier to find and put things together.

I'll read through the rest tomorrow

pitag-ha reviewed (commented):
Here's the second batch of my comments.

Again, I'm super happy with this new manual!

On Monday, I'll finish the review, in particular, I'll have a look at the API changes (btw, just for future reference: it could have been nice to split the PR in two to have the API changes in an isolated PR).

panglesd commented on doc/driver.mld:
```diff
@@ -0,0 +1,372 @@
+transformations from multiple PPXs ensures better composition semantics and
+improves the speed of the combined transformations.
+Moreover, [ppxlib] communicates with Dune through [.corrected] files to allow for
+promotion. Inlining rewriters generate such files, but the PPX author can
```
I'm referring to `deriving_inline` (I think that when I wrote this, I was also incorrectly referring to `Extension.declare_inline`).
But it is very unclear, I'll rewrite this part!!

pitag-ha replied:
Ooh, I see!

panglesd commented on doc/examples.mld:
```diff
@@ -0,0 +1,159 @@
+{%html: <div style="display: flex; justify-content:space-between"><div>%}{{!"good-practices"}< Good practices}{%html: </div><div>%}{%html: </div></div>%}
```
I agree, and I did not removed the example from their folder! But, it is also nice to have full examples accessible in the doc, in my opinion. For example, they will be accessible without leaving ocaml.org!

I will try to make it clearer that the examples are available as "fully working" with dune files etc. with a link to github.

pitag-ha replied:
Yes, I think the examples have a lot of room for improvement on different levels. Sounds good to have them both here and in a folder for now!

panglesd commented on doc/generating-code.mld:
```diff
@@ -0,0 +1,357 @@
+- {!Ppxlib_metaquot}.
+
+{{!Ppxlib.Ast_builder}[Ast_builder]} provides an API to generate AST nodes for the latest OCaml
+version in a backward-compatible way. {!Ppxlib_metaquot} is different: it is a PPX that
+lets you generate OCaml AST nodes by writing OCaml code, using quoting and
+splicing.
```
I'll replace the splicing terms, I read about it on Meta OCaml, but should not have used it!

panglesd commented on doc/generating-code.mld:
```diff
@@ -0,0 +1,357 @@
+
+Using {!Ppxlib_metaquot} requires less knowledge of the OCaml AST than
+{{!Ppxlib.Ast_builder}[Ast_builder]} as it only uses natural OCaml syntax;
+however, it is less general because it only allows users to generate few different nodes such as
+structure items, expressions, patterns, etc., but it is not possible to generate a
+value of type {{!Ppxlib.Parsetree.row_field_desc}[row_field_desc]}!
```
I would not say that it lacks the ability to generate nodes dynamically: antiquotation is a solution for this. The problem is that anti-quotation is a dynamic solution on its own, _but is limited to a few nodes_, so both problems are actually kind of the same!

For instance, the "rec" keyword cannot be generated dynamically: 
```
let is_recursive = ...
let _ =
  [%str let [%a is_recursive] name = ... ]
```
precisely because there is no way to generate just a "rec" value and there is no way to antiquote only the "rec" value.

I'm happy to rephrase, but I'm reluctant to say that "it lacks the ability to generate nodes dynamically" since it would be falsified by the next section!

pitag-ha replied:
> I would not say that it lacks the ability to generate nodes dynamically

What I'm trying to transmit is that `metaquot` _on its own_ isn't able to generate nodes dynamically. For a dynamic node, you need anti-quotation + `Ast_builder` (or another means to build the node).

> I'm happy to rephrase, but I'm reluctant to say that "it lacks the ability to generate nodes dynamically" since it would be falsified by the next section!

You've skipped the "on its own" here :) In the next section, the value you call `dynamic_node` is constructed via `Ast_builder`. So the example doesn't use `metaquot` on its own: the dynamic part is constructed via `Ast_builder` + `mataquot` anti-quotation.

I've tried to make that clear with my last sentence "A typical workflow is to use `metaquot` for the constant skeleton of the node, and to use the `metaquot` anti-quotation workflow (see below) together with `Ast_builder` to fill in the dynamic parts.".

I clearly haven't made clear what I wanted to say, though. Do you agree that it would be nice to add something around this? If not, that's ok with me. If so, let me know if you want me to write a more detailed suggestion!

panglesd replied:
Ok, I think I understand. I was confused by the fact that it is _sometimes_ possible to use metaquot on its own to generate dynamic node:
```
let e = if cond then [%expr 1+1] else [%expr 2+2] in
[%expr 1 + [%e e]
```
However, it is very restricted in what it can generate, and from what data: for instance, it is not possible to create an expression containing a string, given the string itself. Is it what you meant?
Anyway, I agree that it would be nice to add something about this!

```suggestion
however, it's more restrictive than `Ast_builder` for two reasons: first, it's less flexible, since on its own it lacks the ability to generate nodes dynamically from other kind of data: e.g. it's not possible to build an expression containing a string, given the string as input. Second, it's less general because it only allows users to generate few different nodes such as
structure items, expressions, patterns, etc., but it is not possible to generate a
value of type {{!Ppxlib.Parsetree.row_field_desc}[row_field_desc]}! A typical workflow is to use `metaquot` for the constant skeleton of the node, and to use the `metaquot` anti-quotation workflow (see below) together with `Ast_builder` to fill in the dynamic parts.
```

panglesd commented on doc/generating-code.mld:
```diff
@@ -0,0 +1,357 @@
+Yet, every part of a program we write corresponds to a specific AST node, so
+there is no need for AST generation to be more difficult than that.
+
+[Metaquot] is a very useful PPX that allows users to define values of a [Parsetree]
+type by writing natural code, using the quoting and splicing mechanism of
+metaprogramming.
```
I'll use quotation and anti-quotation!

panglesd commented on doc/good-practices.mld:
```diff
@@ -0,0 +1,405 @@
+
+Here is an example on how to write a [debug] rewriter (with the limitation that the payload should not contain variable binding, but the code was left simple to illustrate quoting):
+
+{[
+# let rewrite expr =
+    (* Create a quoter *)
+    let quoter = Quoter.create () in
+    (* An AST mapper to log and replace variables with quoted ones *)
+    let replace_var =
+      object
+        (* See the chapter on AST traverse *)
+        inherit Ast_traverse.map as super
+  
+        (* in case of expression *)
+        method! expression expr =
+          match expr.pexp_desc with
+          (* in case of identifier (not "+") *)
+          | Pexp_ident { txt = Lident var_name; loc }
+            when not (String.equal "+" var_name) ->
+              (* quote the var *)
+              let quoted_var = Quoter.quote quoter expr in
+              let name = Ast_builder.Default.estring ~loc var_name in
+              (* and rewrite the expression *)
+              [%expr
+                debug [%e name] [%e quoted_var];
+                [%e quoted_var]]
+          (* otherwise, continue inside recursively *)
+          | _ -> super#expression expr
+      end
+    in
+    let quoted_rewrite = replace_var#expression expr in
+    let loc = expr.pexp_loc in
+    (* Sanitize the whole thing *)
+    Quoter.sanitize quoter
+      [%expr
+        let debug = Printf.printf "%s = %d; " in
+        [%e quoted_rewrite]] ;;
+  val rewrite : expression -> expression = <fun>
+]}
```
I had trouble finding a simple example where quoting is actually needed...

Moreover, most of the example deals with an AST traverse.

I would be happy if you find a simpler example!

pitag-ha replied:
I'd have to think about it myself as well. We could consider thinking about it together for a short time during hacking days or similar if you want. In any case, for now, I'd leave it like this. It's great to have a section about it!

panglesd replied:
That would be interesting!

panglesd commented on doc/good-practices.mld:
```diff
@@ -0,0 +1,405 @@
+  - : unit = ()
+]}
+
+{1:resp_loc Respecting Locations}
+
+Correctly dealing with location is essential to correctly generate OCaml code.
```
Yes, maybe we should reorganize the order to give more importance to dealing with locations!

However, I don't think we should just say "avoid `Location.none`", but rather "`Location.none` serves this purpose and should only be used in those cases".
For instance, maybe it is better to use `Location.none` rather than having an unrelated location? So that, if the user has an error, at least it knows the location is not relevant?
I would be happy to have some hints on this!

pitag-ha replied:
That's a good point. The only case I can think of where it makes sense to add `Location.none` is when pprinting the generated AST (so that it will get re-parsed by the OCaml compiler with the compiler locations). Btw, that's another section we could add to the manual (for now also as a "not yet written" section. What do you think?

panglesd replied:
After thinking a bit about it, you might be right that there is no situation where it is actually useful: it is only useful when building nodes completely from scratch.
This does happen sometimes in the compiler, but never in ppxlib: at least, we have a file name, and very often more.
So it makes sense to explain why always avoiding Location.none in the ppxlib context.

I am not aware of this mechanism (pprinting-reparsing). I think it could be added as a "to-be-written" section, but hopefully with a small explanation for it not to be too mysterious! (It is a bit mysterious right now for me ;))

panglesd commented on doc/matching-code.mld:
```diff
@@ -0,0 +1,435 @@
+    <fun>
+]}
+
+This function takes a pattern expecting values of type ['matched], continuations of
+type ['cont] and output values of type [('res, _) result] (where the error case is when the ['matched] value does not have the expected structure).
+The types of the function's other arguments correspond to this understanding: the argument of type ['matched] is
+the value to match, the one of type ['cont] is the continuation, and the result
+of applying the pattern to those two values is of type ['res]!
```
I think it is useful to be able to use `parse_res` to match on code. The fact that `Ast_pattern` is mandatory on extenders/derivers, where `parse_res` is called by ppxlib, does not prevent it to be useful in other contexts!

For instance, suppose you traverse some expression and want to extract all sub-expressions with a certain structure. You will use an AST traverse in combination with an `Ast_pattern`, and you will call `parse_res` yourself.

pitag-ha replied:
That's a good point!

panglesd commented on doc/quick_intro.mld:
```diff
@@ -0,0 +1,85 @@
+{{:https://lists.ocaml.org/pipermail/wg-camlp4/2013-January/000000.html}1}
+{{:https://caml.inria.fr/pub/docs/manual-camlp4/manual002.html}2}
+{{:https://camlp5.github.io/}3}
+{{:https://discuss.ocaml.org/t/the-future-of-ppx/3766}4}
```
yes, I had no idea about the fact that the plan was abandoned, and I was very confused while trying to understand "ppxlib's history".

Maybe the best is either to have this section written by someone who know the subject, and won't include random links found on the internet, or remove it altogether (and re-include it later) ?

pitag-ha replied:
So far, the section only includes links, doesn't it?

panglesd replied:
Yes, it only includes links as I could not figure out the whole history, bu tstill wanted to give pointers... But it turned out my pointers were more misleading than useful!
However, those pointers are the one that you find when searching for ppxlib history on the web... so even if we don't include them, people interested in ppxlib history will find them.

What would you prefer:
- Write now (or explain me and I can write) this section
- Remove the confusing links (and add more if you have better pointers
- Remove the section

I don't have a strong opinion!

pitag-ha replied:
All the other links seem good to me in this context. I'd just leave it as is for now (after removing this one link). What do you think?

panglesd replied:
I think that it is a very good option, thanks for deciding!

panglesd commented on doc/writing-ppxs.mld:
```diff
@@ -0,0 +1,519 @@
+  # let extracter () = Ast_pattern.(single_expr_payload (estring __)) ;;
+  val extracter : unit -> (payload, string -> 'a, 'a) Ast_pattern.t = <fun>
+]}
+
+The unit argument is not important.  It is added so that {{:https://v2.ocaml.org/manual/polymorphism.html#ss:valuerestriction}value restriction} does not add noise to the type variables. See the {{!"matching-code"}Destructing AST nodes} chapter and the {{!Ppxlib.Ast_pattern}[Ast_pattern] API} for more explanation on the types and functions used above.
+
```
Do you think it is better to have it here rather than in the "destructing code" chapter?

This chapter is only for the "bureaucracy" to register PPXs. And, it alreay includes "single_expr_payload"!

panglesd replied:
I could add such a section in "destructing code" chapter, and make a direct link from here to there. What do you think?

pitag-ha replied:
I think either would be fine. It will probably mostly be used as a take-it-as-is example, in which case it's part of the boiler-plate. However, people can also take it as a how-to-guide to construct their own pattern. So, maybe the beginning of the destruct section would be a good place. What do you think?

panglesd replied:
I added too many examples to fit them in the "boilerplate", so I added them to the destructing section with a link from the boilerplate to them (for "ready-to-use" patterns).

I prefer to put the example at the end of the destructing section: otherwise, the type does not make sense for someone who has not read the explanation that comes later. Experienced user who only want to quickly refer to examples can use the navigation bar.

panglesd commented on doc/writing-ppxs.mld:
```diff
@@ -0,0 +1,519 @@
+
+{1 Integration with Dune}
+
+If your PPX is written as a Dune project, you'll need to specify the [kind]
+field in your [dune] file with one of the following two values:
+
+- [ppx_rewriter], or
+- [ppx_deriver].
+
+The two possible values exist to allow the coexistence of old [ppx_deriving] derivers and [ppxlib] derivers {e in the toplevel} (see {{:https://github.com/ocaml/dune/issues/1327}this issue} for a discussion on this). On other matter, both stanzas are equivalent. If you are interested in toplevel support, here are the implications of the choice:
+
+- If you use [ppx_rewriter], your rewriter will work out of the box for compilation using [ocamlfind]. As a consequence, it will also work in UTop, which uses [ocamlfind]. However, it won't be possible to use any other [ppx_deriving] deriver in the toplevel.
+- If you use [ppx_deriver], coexistence of your deriver and [ppx_deriving] derivers in UTop will be ensured. However, [ppx_deriving] must become a project dependency to work in the toplevel, even without [ppx_deriving] plugins.
```
Thanks a lot for the explanation!

pitag-ha replied:
I hope this explanation is correct! This is my understanding of it, partly from remembering old conversations with Jeremie and partly from things I've read.

panglesd commented on doc/ast-traversal.mld:
```diff
@@ -0,0 +1,111 @@
+those modules cannot really help. All these examples relate with another kind
+of {{!Ppxlib.Parsetree}[Parsetree]} manipulations known as traversals.
+
+A traversal is a recursive function that will be called on a value, and recursively on all
+of its subvalues, combining the result in a certain way. For instance, {{!Stdlib.List.map}[List.map]} is a traversal of the
+[list] type. In the case of a [list], a map is very simple to write, but in the
+case of the long {{!Ppxlib.Parsetree}[Parsetree]} type, it is a lot of boilerplate code! Fortunately,
+{{!Ppxlib}[ppxlib]} provides a way to ease this.
```
I switched the section on "writing traverses" and the one on the different kind of traverses, to make the example appear early enough so that experienced reader can jump right to it.

Do you think it is sufficient?

pitag-ha replied:
If you think it is, I'm pretty sure it is as well. We can always polish things later if we want to. Thanks for improving it!

pitag-ha commented on doc/driver.mld:
```diff
@@ -0,0 +1,372 @@
+
+{2 The Preprocessing Phase}
+
+The preprocessing phase is the first transformation that actually alters the
+AST. In fact, the property of being the "first transformation applied" is what
+defines this phase, and [ppxlib] will thus ensure that only one transformation is
+registered in this phase; otherwise, it will generate an error.
```
Thanks!

pitag-ha commented on src/ppxlib.ml:
```diff
@@ -1,5 +1,76 @@
+    Opening it will bring several modules in scope, that are useful to have when
+    writing a rewriter:
+
+    - The "core" [ppxlib] modules, such as modules to help manipulate the AST
+      ({!Ast_builder}, {!Ast_pattern}), and a few functions.
```
```suggestion
    - The main [ppxlib] modules, such as modules to help manipulate the AST
      ({!Ast_builder}, {!Ast_pattern}), and a few functions.
```

Just to avoid any wrong association with the Jane Street `core` library. Quite a few people have a strong opinion about avoiding `base` and `core` (and some people associate Ppxlib with them, since we used to have them as dependencies, but we've dropped them some time ago).

pitag-ha commented on src/ppxlib.ml:
```diff
@@ -1,5 +1,76 @@
 
+(** [ppxlib] is supposed to be opened globally in your PPX source files.
+
+    Opening it will bring several modules in scope, that are useful to have when
+    writing a rewriter:
```
```suggestion
    Opening it comes with two advantages. First, it will shadow the [compiler-libs] modules. The [compiler-libs] modules are unstable and aren't meant to be used, so shadowing them is a good protection mechanism. In case you don't want to open [Ppxlib], you can open [Ocaml_shadow] to get the same protection. Second, it will bring several modules in scope, that are useful to have when
    writing a rewriter:
```

pitag-ha commented on src/ppxlib.ml:
```diff
@@ -1,5 +1,76 @@
+(** Include all the Ast definitions since we need them in every single ppx
+
+    @closed *)
+
```
This whole file is a very nice re-ordering! I haven't checked whether some module has gotten lost on the way. Have you double-checked?

panglesd replied:
Yes, I have now double-checked.
Since the compiler-libs include uses destructive substitution, the reordering did not cause any shadowing problem.

pitag-ha reviewed (approved):
This is all awesome! One last time: Thanks a lot!

I'm writing down the first things that have come into my mind that would also be nice mentioning. This is just for future reference, though. I'll make this list more complete and add it to an issue. Let's not add these items now! (or, at most, as "not written yet" sections)
- The hygiene/good practices section could be more complete. For example, we should explain that it's important to fully qualify variable names that are generated into the code via a PPX. In that context, we can also mention and explain runtime modules.
- We could add a small section on how to use the toplevel in the context of PPXs
    - how to load the PPX you're currently developing into `utop`
    - mention `utop-full` and that `Ppxlib` can be used in there (as opposed to `utop` itself)
- As came to my mind in the comment talking about locations, we could add a section on how to use `Ppxlib` without using its driver. For example, how to turn a specification language / something else into OCaml by turning it into an OCaml AST and pprinting it (although we should also point out the limits of `Pprintast` if we do so). 

Just as a side-note: there have been so many comments that I don't have the full overview anymore. If there's any comment that you're expecting me to answer on and I haven't yet, let me know!

pitag-ha commented:
Oh, two more things:

We need to cut a release for this to be generated, right? Do you want to take care of that?

And: I very strongly encourage you to write about this on discuss once it's live!

panglesd commented on src/common.mli:
```diff
@@ -1,3 +1,5 @@
+(** Common functions *)
+
```
Normally, doc comments are kept when included in another signature.

However, this particular one seems to be dropped for some reasons (maybe https://github.com/ocaml/odoc/issues/878?)

panglesd commented:
Thanks a lot @pitag-ha for the review!

I have answered and got answered on all comments, it is ready to be merged on my side. (Do you want me to squash all commits?)

> I'm writing down the first things that have come into my mind that would also be nice mentioning. This is just for future reference, though. I'll make this list more complete and add it to an issue. Let's not add these items now! (or, at most, as "not written yet" sections)

I totally agree with this strategy (and with the items themselves)!

> We need to cut a release for this to be generated, right? Do you want to take care of that?

Yes, I can do a release!

> And: I very strongly encourage you to write about this on discuss once it's live!

Sure, I will do! I hope it might encourage other important OCaml projects to write more comprehensive manuals.

pitag-ha commented:
Perfect! Yes, squashing sounds good.

Btw, one more point that's come to my mind (just like the others: we shoudln't add it _now_): You've already mentioned stability throughout the document. Additional to that, we could also add a dedicated section about stability. Mostly for two things: to sum up how to make the PPX the most stable possible; and to explain how we're keeping the PPX ecosystem up-to-date and inter-compatible (i.e. the `ppx-universe` and patch PRs workflow).

tatchi commented on doc/generating-code.mld:
```diff
@@ -0,0 +1,360 @@
+First, in order to use [Metaquot], add it in your [preprocess] Dune stanza:
+
+{[
+  (prepocess (pps ppxlib.metaquot))
```
```suggestion
  (preprocess (pps ppxlib.metaquot))
```

panglesd replied:
Thanks!
